### ISSUE TITLE
Reproducing end-to-end training numbers from FA2 paper

### ISSUE THREAD

**Issue by zplizzi** [Dao-AILab/flash-attention#558]
I'm looking to reproduce this result:

> Overall this speeds up training by 3-5x compared to the baseline implementation from Huggingface, reaching up to 225 TFLOPs/sec per A100, equivalent to 72% model FLOPs utilization (we don't need any activation checkpointing).

But I don't see in the paper or in this repository where it's documented exactly what config/code was used to generate that result. The training code and documentation in https://github.com/Dao-AILab/flash-attention/tree/main/training looks to be mostly from flash-attn 1, and the readme in that folder looks to report lower/older performance numbers.

Could you point me to the config/command to run where I'd expect to see similar performance numbers? Should I just follow the old training instructions, but with FA2, and expect to achieve those new numbers? Thanks!

**Comment by tridao** [comment]
Yeah it should be the same training code, since it calls to `gpt.py` and `mha.py` which has the new implementation. Just the README hasn't been updated.

**Comment by zplizzi** [comment]
Thanks!

**Comment by zplizzi** [comment]
How did you handle GPU clocks and throttling when benchmarking and reporting these figures? On the machine I'm benchmarking on, my A100s drop well below the peak boost clock within seconds of running a workload. Is the MFU calculation still just ("effective steady-state flops" / "312 ideal boost TFLOPS"), or do you lock clocks to a lower value and derate the denominator (ideal TFLOPs) also?

**Comment by tridao** [comment]
I measured the state-state flops and divide by 312 TFLOPs.

**Comment by zplizzi** [comment]
I'm not sure - locking the clocks is helpful for reproducibility, as it makes it easier to control for differences in chip lottery or machine cooling or whatnot. But it depends on if you're trying to measure just software efficiency (in which case locking the clocks and potentially derating the max FLOPs number also might make sense), or whole system efficiency (in which case just leaving things stock probably makes sense).

Regardless, would you mind sharing a bit more details of how you came to that 72% MFU figure? I'm trying to benchmark some of your training code, but it doesn't seem like MFU is actually reported in the logs. I see code for FLOP counting (which I can't get to run readily), and for speed monitoring (are you using `intra_step_time` or `intra_step_time + inter_step_time` in the MFU calculation?). I'm also not sure which model size you were using. If I substitute in my best  guess for some of these things, use an approximation for model FLOPs, and run on one 80GB A100 with a ~750M param model, device batch size 8 and sequence length 2048, I'm getting a MFU of around 61% (with my graphics clocks unlocked and hovering around 1300MHz steady-state).

Maybe you could share the training config and some timings (or the wandb logs) from that benchmark run, so I can see if I can reproduce the speed - to separate out the nuances in computing MFU from how fast it actually runs.

**Comment by tridao** [comment]
This was with 2.7B on 8k context size as mentioned in the technical [report](https://arxiv.org/abs/2307.08691). I think it's this [config](https://github.com/Dao-AILab/flash-attention/blob/main/training/configs/experiment/pile/gpt3-2.7B-flash-hdim128-rotary-8k.yaml), with dropout disabled.
`inter_step_time` is pretty small so you can use either `intra_step_time` or `intra_step_time + inter_step_time`.

**Comment by zplizzi** [comment]
Gotcha, thanks, I had missed table 1 in that report. From that table, a 1.3B model with 2k sequence length gets 196/312=63% MFU, so the numbers I reported above are not too far off.
